Custom JS Libraries (#719)
appsmithorg/appsmith-docs: website/docs/core-concepts/writing-code/ext-libraries.md

# External Libraries

With Appsmith, you have the ability to install and utilize external JavaScript libraries to enhance your ability to manipulate data within `{{ }}` bindings or within JSObjects. These external libraries can provide you with additional methods to help you build complex application and business logic.

>  <VideoEmbed host="youtube" videoId="tqJna718tj4" title="Using Built-in Libraries in Appsmith" caption="Using Built-in Libraries in Appsmith"/> 

Author: done

> ## Pre-installed JavaScript libraries

Author: Should we use the word "built-in"? As they are external libraries, which are pre-installed in appsmith @jnikhila your thoughts?

There are variety of pre-installed JavaScript libraries that can be utilized in your projects. These libraries provide access to a range of useful functions and can be incorporated into your applications.

> 

Author: done

* [Lodash](https://lodash.com/docs/4.17.15), is a JavaScript utility library that provides functions for common programming tasks such as formatting data, iterating over collections, and manipulating arrays and objects.

* [Moment](https://momentjs.com/docs/), is a JavaScript library that simplifies working with dates and times in JavaScript by providing functions for parsing, validating, manipulating, and displaying dates and times.

* [xmlParser](https://naturalintelligence.github.io/fast-xml-parser/), is a library for parsing and manipulating XML data in JavaScript.

* [Forge](https://github.com/digitalbazaar/forge), is a library for working with cryptographic algorithms and protocols in JavaScript.

## Install custom JavaScript libraries

You can browse and install recommended JS libraries or install a JS library of your choice by providing a valid URL. 

For example- 

> *  Valid and supported:  `https://cdn.jsdelivr.net/npm/exceljs@4.3.0/dist/exceljs.min.js`. 

Author: updated

* Invalid URL: `https://www.jsdelivr.com/package/npm/datejs`

* Valid URL but not supported: `https://cdn.jsdelivr.net/npm/uuid@9.0.0/dist/index.js`

> You can search for libraries on popular CDN services like [jsDelivr](https://www.jsdelivr.com/) or [UNPKG](https://unpkg.com/) to find the URL, which points to the library's index file. It's important to ensure that the library supports a [UMD](https://github.com/umdjs/umd) build to work with Appsmith. For more information, see the [basic pattern](https://github.com/umdjs/umd/blob/master/templates/commonjsStrict.js) of a UMD build.

Author: removed

> Many libraries may have a `.min.js` file located in the `root`, `/umd`, or `/browser` folders. If the library you wish to use doesn't support a UMD build, you can use `browserify` to create one and host it on a CDN of your choice.

Author: removed

> 

Reviewer: Let there be a subheading under this called **Install external library**

Author: added a sub head

> Follow the below steps to install a custom JavaScript library:

Reviewer: we can delete this line

Author: updated the section intro

> ![Installing custom JavaScript libraries](/img/customjs.gif)

Reviewer: the resolution is really bad. I thought we are moving away from gifs and going to do videos instead. Can we do that and show both how to install recommended library and also using the url

Author: video added

> * On the Explorer tab,

Reviewer: ```suggestion
* Navigate to the Explorer tab
```

* Click the `+` sign next to `Libraries`.

> * Paste the URL into the designated field. For example: 

Reviewer: Above this point, also mention that they can click the icon to install the recommended libraries. and then go on to mention how they can paste in the URL for other libraries.

Author: done

```js 
https://cdn.jsdelivr.net/npm/exceljs@4.3.0/dist/exceljs.min.js
```
* Click `Install`.

> ## Using JavaScript libraries

Reviewer: ```suggestion
## Using external libraries
```

* [Lodash](https://lodash.com/docs/4.17.15), is a JavaScript utility library that provides functions for common programming tasks such as formatting data, iterating over collections, and manipulating arrays and objects.

* [Moment](https://momentjs.com/docs/), is a JavaScript library that simplifies working with dates and times in JavaScript by providing functions for parsing, validating, manipulating, and displaying dates and times.

* [xmlParser](https://naturalintelligence.github.io/fast-xml-parser/), is a library for parsing and manipulating XML data in JavaScript.

* [Forge](https://github.com/digitalbazaar/forge), is a library for working with cryptographic algorithms and protocols in JavaScript.

## Installing custom JavaScript libraries

Browse and install recommended JS libraries or install a JS library of your choice by pasting a valid URL. You can search libraries on popular CDN services like [jsDelivr](https://www.jsdelivr.com/) or [UNPKG](https://unpkg.com/) and obtain the URL. 

> ![](/img/customjs.gif)

Reviewer 2: Please add a caption.

Author: done

To install a custom JavaScript library, 

* On the Explorer tab,
* Click the `+` sign next to `Libraries`.
* Paste the URL into the designated field. For example: 
```js 
https://cdn.jsdelivr.net/npm/exceljs@4.3.0/dist/exceljs.min.js
```
* Click `Install`.

:::info
Please use a URL that points to the library's index file. Also ensure that your library supports a **[UMD](https://github.com/umdjs/umd)** build for it to work on Appsmith. Here’s the [basic pattern](https://github.com/umdjs/umd/blob/master/templates/commonjsStrict.js) of a UMD build. Most libraries might have a `.min.js` under the `root`, `/umd` or `/browser` folders. If a library that you wish to use doesn't support a UMD build, you may use [browserify](https://browserify.org/) to generate one and host it in a CDN of your choice. 

❌ Invalid URL -> https://www.jsdelivr.com/package/npm/datejs

❌ Valid URL but unsupported library module format -> https://cdn.jsdelivr.net/npm/uuid@9.0.0/dist/index.js

✅ Valid URL and supported library module format -> https://cdn.jsdelivr.net/npm/exceljs@4.3.0/dist/exceljs.min.js
:::

## Using JavaScript libraries

> The external libraries can be used anywhere inside `{{ }}` just as JavaScript is used in the rest of the application. The signature of the JavaScript libraries are exactly the same as mentioned in their documentation. 

Reviewer 2: ```suggestion
External libraries can be used within `{{ }}` in the same way JavaScript is used elsewhere in the application. For more information about the signature of the JavaScript libraries, see their official documentation.
```

# External Libraries

The Appsmith platform includes built-in JavaScript utility libraries, which can be used to work with data within `{{ }}` bindings or within JSObjects. You can also install and use other custom libraries to manipulate or transform data. These external libraries provide additional methods to help you build complex applications and business logic.

## Built-in JavaScript libraries

> Appsmith provides the following built-in JavaScript libraries that can be utilized in your applications.

Author: But only heading and video will look odd. We might need a small intro between the heading and the video. @jnikhila thoughts?

> Appsmith provides the following built-in JavaScript libraries that can be utilized in your applications.

Author: @danciaclara added another intro and placed the line below the video. have a look and let me know.

* [xmlParser](https://naturalintelligence.github.io/fast-xml-parser/) can be used for parsing and manipulating XML data in JavaScript.

* [Forge](https://github.com/digitalbazaar/forge) can be used to work with cryptographic algorithms and protocols in JavaScript.

## Custom JavaScript libraries

Browse and install recommended JS libraries or install a JS library of your choice by pasting a valid URL. You can search libraries on popular CDN services like [jsDelivr](https://www.jsdelivr.com/) or [UNPKG](https://unpkg.com/) to obtain the URL.

Use a URL that points to the library's index file. Ensure that your library supports a **[UMD](https://github.com/umdjs/umd)** build for it to work on Appsmith. Here’s the [basic pattern](https://github.com/umdjs/umd/blob/master/templates/commonjsStrict.js) of a UMD build. Most libraries have a `.min.js` under the `root`, `/umd` or `/browser` folders. If a library you wish to use doesn't support a UMD build, you may use [browserify](https://browserify.org/) to generate one and host it in a CDN of your choice.

> 

Author: done

}}
```

#### moment

An example of the moment.js `format` utility, in use in a [query](../data-access-and-binding/querying-a-database/).

```sql
insert into users (name, email, createdDate) values 
('John', 'john@appsmith.com', {{moment().format("YYYY-MM-DD")}})
```
#### excelJS

You can install excelJS library using this [URL](https://www.jsdelivr.com/package/npm/exceljs) .

```javascript
createWorkbook: async() => {
		const workbook = new ExcelJS.Workbook();
    workbook.creator = 'Tomato';
    workbook.lastModifiedBy = 'Tomato';
    workbook.created = new Date();
    workbook.modified = new Date();
    workbook.calcProperties.fullCalcOnLoad = true;

>     const worksheet = workbook.addWorksheet('Tomato page 1', {

Reviewer 3: could we add a line break here for better readability?

      properties: {tabColor: {argb: '#FF0000'}},
      pageSetup: {paperSize: 9, orientation: 'landscape'}
    });
		worksheet.getCell('A1').value = 'Hello, World!';
    worksheet.getCell('B1').value = 'What time is it?';
    worksheet.getCell('A2').value = 7;
    worksheet.getCell('B2').value = '12pm';

>     const buf = await workbook.xlsx.writeBuffer();

Reviewer 3: ++ line break

	  const blob = new Blob([buf], {type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'});
	  const url = URL.createObjectURL(blob);
    await download(
			url, 
			"test.xls",
		);

> 	}

Reviewer 3: indentation needs to be fixed

```sql
insert into users (name, email, createdDate) values 
('John', 'john@appsmith.com', {{moment().format("YYYY-MM-DD")}})
```
#### excelJS

You can install excelJS library using this [URL](https://www.jsdelivr.com/package/npm/exceljs) .

```javascript
createWorkbook: async() => {
		const workbook = new ExcelJS.Workbook();
    workbook.creator = 'Tomato';
    workbook.lastModifiedBy = 'Tomato';
    workbook.created = new Date();
    workbook.modified = new Date();
    workbook.calcProperties.fullCalcOnLoad = true;
    const worksheet = workbook.addWorksheet('Tomato page 1', {
      properties: {tabColor: {argb: '#FF0000'}},
      pageSetup: {paperSize: 9, orientation: 'landscape'}
    });
		worksheet.getCell('A1').value = 'Hello, World!';
    worksheet.getCell('B1').value = 'What time is it?';
    worksheet.getCell('A2').value = 7;
    worksheet.getCell('B2').value = '12pm';
    const buf = await workbook.xlsx.writeBuffer();
	  const blob = new Blob([buf], {type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'});
	  const url = URL.createObjectURL(blob);
    await download(
			url, 
			"test.xls",
		);
	}
```
## General notes

> Please note that you may not be able to install or use methods of certain libraries due to platform limitations:

Reviewer 2: ```suggestion
You may not be able to install or use methods of certain libraries due to platform limitations:
```

* [DOM](https://developer.mozilla.org/en-US/docs/Web/API/Document_Object_Model/Introduction) access: Libraries that try to manipulate document objects won’t work. Example: https://d3js.org/
* [XHR](https://www.notion.so/Custom-JS-Libraries-82c03d95918b4eaa8f3e0dd811f3cd00): Libraries that only rely on XHR won’t work.
* Other APIs: Library methods that use the following APIs under the hood won’t work: [setInterval](https://developer.mozilla.org/en-US/docs/Web/API/setInterval), [clearInterval](https://developer.mozilla.org/en-US/docs/Web/API/clearInterval), [setImmediate](https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate), [localStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) and [Navigator](https://developer.mozilla.org/en-US/docs/Web/API/Navigator).

### Troubleshooting

> If you are having trouble connecting External Libraries, check out this [troubleshooting guide](help-and-support/troubleshooting-guide/js-errors) or reach out on [Discord](https://discord.com/invite/rBTTVJp) to get support or ask questions on the [community forum](https://community.appsmith.com/).

Reviewer 2: ```suggestion
If you are experiencing difficulties with connecting external libraries, you can refer to the [JavaScript Errors troubleshooting guide](help-and-support/troubleshooting-guide/js-errors) for assistance or seek help on [Discord](https://discord.com/invite/rBTTVJp). You can also ask questions and get support on the [community forum](https://community.appsmith.com/).
```
